### ISSUE TITLE
[New rule] avoid-dynamic

### ISSUE THREAD

**Issue by roman-petrov** [dart-code-checker/dart-code-metrics#472]
**Please describe what the rule should do:**

The rule should entirely prohibit using the `dynamic` keyword. There are plenty of standard Dart linter rules disallowing `dynamic`, but it seems they are not strict enough. `dynamic` is usually intended to be used for dynamic data, like JSON. Using `dynamic` without intention in strictly typed code might lead to runtime errors. I am starting to see this quite often in PRs from inexperienced developers.

For cases where we **really** need dynamic, we can disable this rule in isolated code which uses dynamic data. Also, we can use new type aliases feature like `typedef Json = Map<String, dynamic>;`, as shown in [Dart 2.13 announcement](https://medium.com/dartlang/announcing-dart-2-13-c6d547b57067)

**If your rule is inspired by other please provide link to it:**

* TypeScript ESLint [no-explicit-any](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-explicit-any.md)
* TSLint [no-any](https://palantir.github.io/tslint/rules/no-any/)

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**

Bad:
```dart
dynamic x = 10;
String concat(dynamic a, dynamic b);
```

Good:
```dart
int x = 10;
final x = 10;
String concat(String a, String b);
```

**Are you willing to submit a pull request to implement this rule?**
Maybe

**Comment by incendial** [comment]
> There are plenty of standard Dart linter rules disallowing dynamic, but it seems they are not strict enough. 

@roman-petrov could you please provide some examples of these rules and the cases that are missed?

**Comment by roman-petrov** [comment]
Actually, after looking into all Dart rules I see just one rule which protects from calling dynamics, [avoid_dynamic_calls](https://dart-lang.github.io/linter/lints/avoid_dynamic_calls.html).

So I do not see any way how to prohibit codes like

```dart
String concat(dynamic a, dynamic b) {
    if (a is String && b is String) {
        return [a, b].join(' ');
    }

   return '';
}
```

This is almost a real code I found when reviewing code coming from an inexperienced developer. I do not know why some people write such code, but at the moment the only way to disallow such things for me is a manual complex review process.

That's why I am proposing a rule to completely disallow the `dynamic` type. Maybe, I am missing something here, please let me know if there is a way to disallow such things in Dart.

**Comment by incendial** [comment]
What about this one https://dart-lang.github.io/linter/lints/avoid_annotating_with_dynamic.html ?

**Comment by roman-petrov** [comment]
So is now the idea clear, do you think this rule would be valuable?..

If so - lets have it opened, maybe I or someone else will make a PR for it.

**Comment by incendial** [comment]
> So is now the idea clear, do you think this rule would be valuable?..

Yes, I think it's pretty solid especially for the Dart newcomers.

**Comment by incendial** [comment]
@roman-petrov seems like there is a new rule https://dart-lang.github.io/linter/lints/avoid_dynamic_calls.html, is it what you wanted?

**Comment by roman-petrov** [comment]
@incendial, `avoid-dynamic-calls` as I know, is actually pretty old rule. And no - this is not the rule I wanted: it disallows only dynamic calls as shown in the docs.

My idea is to totally disallow using `dynamic` keyword. In regular Dart / Flutter code there is usually no need to use dynamic. `dynamic` keyword is usually being used in some interop code when manipulating with "dynamic" data: JSON, FFI, JavaScript objects, etc. But in a good architecture these code parts usually isolated from main business logic.

The original reason I created this issue was a function I seen in a PR like:

```dart
String concat(dynamic firstName, dynamic lastName) {
    if (firstName is String && lastName is String) {
        return [firstName, lastName].join(' ');
    }

   return '';
}
```

I know that this function looks pretty strange and usually developers don't write such functions :) But I think this is a quite bad implementation and it would be good to avoid such patterns using linter.

I don't think that this rule will give very much value, but it still can be useful.

**Comment by incendial** [comment]
> I don't think that this rule will give very much value, but it still can be useful.

I disagree, I think this check is pretty decent, but I still trying to understand its proper scope and that's why there are additional questions and other rule examples. But it looks like I finally get it. Thank you!

**Comment by dkrutskikh** [comment]
Available in 4.11.0 release 🚀
